core/src/main/scala/kafka/log/LogCleaner.scala

```diff
@@ -307,24 +332,39 @@ class LogCleaner(initialConfig: CleanerConfig,
             case _: LogCleaningAbortedException => // task can be aborted, let it go.
             case _: KafkaStorageException => // partition is already offline. let it go.
             case e: IOException =>
-              val msg = s"Failed to clean up log for ${cleanable.topicPartition} in dir ${cleanable.log.dir.getParent} due to IOException"
-              logDirFailureChannel.maybeAddOfflineLogDir(cleanable.log.dir.getParent, msg, e)
+              var logDirectory = cleanable.log.dir.getParent
+              val msg = s"Failed to clean up log for ${cleanable.topicPartition} in dir ${logDirectory} due to IOException"
+              cleanerManager.markDirectoryUncleanable(logDirectory)
+              logDirFailureChannel.maybeAddOfflineLogDir(logDirectory, msg, Some(e))
+            case e: Exception => throw e
           } finally {
             cleanerManager.doneCleaning(cleanable.topicPartition, cleanable.log.dir.getParentFile, endOffset)
           }
           true
       }
       val deletable: Iterable[(TopicPartition, Log)] = cleanerManager.deletableLogs()
-      deletable.foreach{
+      deletable.foreach {
         case (topicPartition, log) =>
           try {
+            currentLog = Some(log)
             log.deleteOldSegments()
-          } finally {
+          } catch {
+            case e: KafkaStorageException =>
+              e.getCause match {
+                case exception: IOException =>
+                  val logDirectory = log.dir.getParent
+                  val msg = s"Failed to delete log for segment ${log.topicPartition} in dir ${logDirectory} due to IOException"
+                  cleanerManager.markDirectoryUncleanable(logDirectory)
+                  logDirFailureChannel.maybeAddOfflineLogDir(logDirectory, msg, Some(exception))
+              }
+              throw e
+          } finally  {
```

Comment:
I don't think we do. The idea is to have `logBeingCleaned` (previously `currentLog`) be populated if an exception is raised. That way `cleanOrSleep()` knows which was the faulty log. 
I think it's clearer to make the variable `None` in the same method it is assigned.

After this comment, I saw that if `cleanerManager#grabFielthiestCompactedLog()` raises an exception, we'll re-raise it as an `IllegalStateException`. Not sure how favorable that is, since that would kill the thread. There still exist code paths that can kill the thread outright (via the re-raising of an exception), but I'm not sure how else to handle such scenarios. Maybe this is okay